### PR TITLE
http_health_check: host and path validation

### PR DESCRIPTION
**Pull request by botengyao**
The previous `host` and `path` value cannot pass the encodeHeader validation.
Added stricter validation for both.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/config/core/v3/health_check.proto [hunk 1]
```diff
@@ -95,12 +95,12 @@ message HealthCheck {
     // left empty (default value), the name of the cluster this health check is associated
     // with will be used. The host header can be customized for a specific endpoint by setting the
     // :ref:`hostname <envoy_v3_api_field_config.endpoint.v3.Endpoint.HealthCheckConfig.hostname>` field.
-    string host = 1 [(validate.rules).string = {well_known_regex: HTTP_HEADER_VALUE strict: false}];
+    string host = 1 [(validate.rules).string = {well_known_regex: HTTP_HEADER_VALUE}];
```
[the remark this answers is not in the document]
Comment (botengyao) [reply api/envoy/config/core/v3/health_check.proto]: Thanks & PTAL, I tend to use the same level validation as [checkValidRequestHeaders](https://github.com/envoyproxy/envoy/blob/9a47bc9eff9c3283fbfaed9144313b7bda366aac/source/common/http/http1/codec_impl.cc#L449)

File: source/extensions/health_checkers/http/health_checker_impl.cc [hunk 2]
```diff
@@ -81,6 +81,12 @@ HttpHealthCheckerImpl::HttpHealthCheckerImpl(const Cluster& cluster,
     service_name_matcher_.emplace(config.http_health_check().service_name_matcher());
   }
 
+  if (!Http::HeaderUtility::headerValueIsValid(path_) ||
+      !Http::HeaderUtility::headerValueIsValid(host_value_)) {
+    throw EnvoyException(fmt::format("the path {} or host name {} is not a valid header value.",
```
Comment (adisuissa) [inline source/extensions/health_checkers/http/health_checker_impl.cc:86]: nit: prefer separating to different if statements, and giving an explicit error to the field that causes the problem.
Also instead of throw, please use `THROW_IF_STATUS_NOT_OK/THROW_IF_NOT_OK`.

File: source/extensions/health_checkers/http/health_checker_impl.cc [hunk 3]
```diff
@@ -81,6 +81,12 @@ HttpHealthCheckerImpl::HttpHealthCheckerImpl(const Cluster& cluster,
     service_name_matcher_.emplace(config.http_health_check().service_name_matcher());
   }
 
+  if (!Http::HeaderUtility::headerValueIsValid(path_) ||
+      !Http::HeaderUtility::headerValueIsValid(host_value_)) {
```
Comment (adisuissa) [inline source/extensions/health_checkers/http/health_checker_impl.cc:85]: For host consider using: `HeaderUtility::authorityIsValid`.

File: source/extensions/health_checkers/http/health_checker_impl.cc [hunk 4]
```diff
@@ -81,6 +81,12 @@ HttpHealthCheckerImpl::HttpHealthCheckerImpl(const Cluster& cluster,
     service_name_matcher_.emplace(config.http_health_check().service_name_matcher());
   }
 
+  if (!Http::HeaderUtility::headerValueIsValid(path_) ||
```
Comment (adisuissa) [inline source/extensions/health_checkers/http/health_checker_impl.cc:84]: For path validation, and because it seems `headerValueIsValid` is using QUICHE's function, why not use the [`IsValidPath` function](https://github.com/google/quiche/blob/a4c007178d479e7cc301a2bae284f41c87039448/quiche/http2/adapter/header_validator.h#L44) in QUICHE?

File: source/extensions/health_checkers/http/health_checker_impl.cc (hunk 3 above)
Comment (adisuissa) [reply source/extensions/health_checkers/http/health_checker_impl.cc:85]: Why? Please add a comment with the reason.

Comment (botengyao) [reply source/extensions/health_checkers/http/health_checker_impl.cc]: The initial thought is to keep the same validation as the beginning of `encodeHeaders()`, but I don't have objections to separate them. Done  & PTAL

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!
/assign-from @envoyproxy/senior-maintainers

**Comment by adisuissa** [comment]
Not sure if it requires a release note or not, and will leave the decision to the senior maintainer.

File: source/extensions/health_checkers/http/health_checker_impl.cc [hunk 5]
```diff
@@ -81,15 +81,25 @@ HttpHealthCheckerImpl::HttpHealthCheckerImpl(const Cluster& cluster,
     service_name_matcher_.emplace(config.http_health_check().service_name_matcher());
   }
 
+  if (!Http::HeaderUtility::authorityIsValid(host_value_)) {
```
Comment (htuch) [inline source/extensions/health_checkers/http/health_checker_impl.cc:84]: @yanavlasov how do HCs fit with UHV?

**Comment by KBaichoo** [comment]
friendly ping @yanavlasov

File: source/extensions/health_checkers/http/health_checker_impl.cc (hunk 5 above)
Comment (yanavlasov) [reply source/extensions/health_checkers/http/health_checker_impl.cc:84]: We have discussed with @botengyao . Presently there is no way to use UHV since it is still behind `#ifdef`, so just adding PGV validator to avoid ASSERTs in the fuzzer should be sufficient.

@botengyao please add a TODO here to introduce additional validation for the authority and path headers based on the default UHV when it is available.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/health_checkers/http/health_checker_impl.cc (hunk 5 above)
Comment (botengyao) [reply source/extensions/health_checkers/http/health_checker_impl.cc:86]: Done, thanks!

Comment (yanavlasov) [reply source/extensions/health_checkers/http/health_checker_impl.cc:86]: I think you can just PGV annotations to the config proto and avoid these custom checks here altogether.

**Review by yanavlasov — COMMENTED** [review]
/wait-any

File: source/extensions/health_checkers/http/health_checker_impl.cc (hunk 5 above)
Comment (botengyao) [reply source/extensions/health_checkers/http/health_checker_impl.cc:86]: Thanks for the review!

Per discussion with @adisuissa  we want some more clarity on whether a stricter host and path validations are needed in this case, decided to use this customized logic. 

Here is the PGV set which is only header value check: https://github.com/bufbuild/protoc-gen-validate/blob/7d833c1f32da0f46ebecc1f3f88cf093cd73a6e6/python/protoc_gen_validate/validator.py#L27

There are some small difference for path and header:

```
constexpr absl::string_view kValidAuthorityChars =
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~%!$&'()["
    "]*+,;=:";

constexpr absl::string_view kValidPathChars =
    "/abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-._~%!$&'()"
    "*+,;=:@?";
```

I am fine to directly use the HTTP_HEADER_VALUE PGV check, but current check is stricter.

Comment (botengyao) [reply source/extensions/health_checkers/http/health_checker_impl.cc]: Discussed offline, and will use PGV right now till UHV is available.

File: source/extensions/health_checkers/http/health_checker_impl.cc [hunk 6]
```diff
@@ -86,9 +88,9 @@ HttpHealthCheckerImpl::HttpHealthCheckerImpl(
       total += bytes.size();
     }
     if (total > response_buffer_size_) {
-      throw EnvoyException(fmt::format(
```
Comment (yanavlasov) [inline source/extensions/health_checkers/http/health_checker_impl.cc:89]: Is this change intentional?

**Review by yanavlasov — COMMENTED** [review]
LGTM with a question.

/wait-any

File: source/extensions/health_checkers/http/health_checker_impl.cc (hunk 6 above)
Comment (botengyao) [reply source/extensions/health_checkers/http/health_checker_impl.cc:89]: Yes, trying to avoid direct throw and replace with the macro.

Comment (yanavlasov) [reply source/extensions/health_checkers/http/health_checker_impl.cc:89]: We use throw elsewhere in this file. The change looks out of scope for what you need to accomplish. I would just omit it.

Comment (botengyao) [reply source/extensions/health_checkers/http/health_checker_impl.cc]: done

**Comment by yanavlasov** [comment]
LGTM.
Please merge main to resolve the coverage error.
